src/Avalonia.Controls/Presenters/TextPresenter.cs

```diff
@@ -443,113 +447,316 @@ internal void CaretIndexChanged(int caretIndex)
+                    new ValueSpan<TextRunProperties>(start, length,
+                        new GenericTextRunProperties(typeface, FontSize,
+                            foregroundBrush: SelectionForegroundBrush ?? Brushes.White))
                 };
             }
 
+            if (PasswordChar != default(char) && !RevealPassword)
+            {
+                result = CreateTextLayoutInternal(_constraint, new string(PasswordChar, text?.Length ?? 0), typeface,
+                    textStyleOverrides);
+            }
+            else
+            {
+                result = CreateTextLayoutInternal(_constraint, text, typeface, textStyleOverrides);
+            }
+
             return result;
         }
 
-        /// <summary>
-        /// Measures the control.
-        /// </summary>
-        /// <param name="availableSize">The available size for the control.</param>
-        /// <returns>The desired size.</returns>
-        private Size MeasureInternal(Size availableSize)
+        protected virtual void InvalidateTextLayout()
+        {
+            _textLayout = null;
+            
+            InvalidateMeasure();
+        }
+        
+        protected override Size MeasureOverride(Size availableSize)
         {
-            if (!string.IsNullOrEmpty(Text))
+            if (!double.IsInfinity(availableSize.Width) && availableSize != _constraint)
+            {
+                _constraint = availableSize;
+                
+                InvalidateTextLayout();
+            }
+
+            return TextLayout.Size;
+        }
+
+        private int CoerceCaretIndex(int value)
+        {
+            var text = Text;
+            var length = text?.Length ?? 0;
+            return Math.Max(0, Math.Min(length, value));
+        }
+
+        private void CaretTimerTick(object sender, EventArgs e)
+        {
+            _caretBlink = !_caretBlink;
+            InvalidateVisual();
+        }
+
+        public void MoveCaretToTextPosition(int textPosition, bool trailingEdge = false)
+        {
+            var lineIndex = TextLayout.GetLineIndexFromCharacterIndex(textPosition, trailingEdge);
+            var textLine = TextLayout.TextLines[lineIndex];
+
+            var characterHit = textLine.GetPreviousCaretCharacterHit(new CharacterHit(textPosition));
+            
+            var nextCaretCharacterHit = textLine.GetNextCaretCharacterHit(characterHit);
+
+            if (nextCaretCharacterHit.FirstCharacterIndex <= textPosition)
+            {
+                characterHit = nextCaretCharacterHit;
+            }
+
+            if (textPosition == characterHit.FirstCharacterIndex + characterHit.TrailingLength)
+            {
+                UpdateCaret(characterHit);
+            }
+            else
             {
-                if (TextWrapping == TextWrapping.Wrap)
+                UpdateCaret(trailingEdge ? characterHit : new CharacterHit(characterHit.FirstCharacterIndex));
+            }
```

Comment:
This might need to set `_navigationPosition` since other callsites seem do do it?
In my tests after changing this then navigation after text input works fine.